libraries/chain/db_market.cpp

```diff
@@ -842,33 +840,30 @@ bool database::fill_call_order( const call_order_object& order, const asset& pay
             else if( get_dynamic_global_properties().next_maintenance_time > HARDFORK_CORE_343_TIME )
               o.call_price = price::call_price( o.get_debt(), o.get_collateral(),
                                 mia.bitasset_data(*this).current_feed.maintenance_collateral_ratio );
-       });
+      });
 
+   // update current supply
    const asset_dynamic_data_object& mia_ddo = mia.dynamic_asset_data_id(*this);
 
-   modify( mia_ddo, [&]( asset_dynamic_data_object& ao ){
-       //idump((receives));
-        ao.current_supply -= receives.amount;
+   modify( mia_ddo, [&receives]( asset_dynamic_data_object& ao ){
+         ao.current_supply -= receives.amount;
       });
 
-   const account_object& borrower = order.borrower(*this);
+   // adjust balance and/or account statistics
    if( collateral_freed.valid() || pays.asset_id == asset_id_type() )
    {
-      const account_statistics_object& borrower_statistics = borrower.statistics(*this);
       if( collateral_freed.valid() )
-         adjust_balance(borrower.get_id(), *collateral_freed);
+         adjust_balance( order.borrower, *collateral_freed );
 
-      modify( borrower_statistics, [&]( account_statistics_object& b ){
-              if( collateral_freed.valid() && collateral_freed->amount > 0 && collateral_freed->asset_id == asset_id_type() )
-                b.total_core_in_orders -= collateral_freed->amount;
+      modify( get_account_stats_by_owner(order.borrower), [&collateral_freed,&pays]( account_statistics_object& b ){
```

Comment:
Only call modify if ``pays.asset_id == asset_id_type()`` (this also implies collateral_freed.asset_id == core if collateral_freed is defined)